### PR TITLE
Add Bosch BMM350 magnetometer

### PR DESCRIPTION
**Pull request by Viliuks**
### Solved Problem
Added Bosch BMM350 magnetometer released in 2023 with variable sampling, averaging rates. And another option to set the pad drive settings, to fix issues with data overshooting/undershooting that happen using long wires.

New parameters:
```
BMM350_ODR - sets ODR mode
BMM350_AVG - sets averaging mode
BMM350_DRIVE - sets pad drive mode
```
### Context
https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bmm350-ds001.pdf

### CODE REVIEW

**Comment by PetervdPerk-NXP** [comment]
@jlaitine You made a BMM350 driver as well right for PX4?
Could you review this, I don't have hardware yet with a BMM350 but we're planning to.

**Comment by jlaitine** [comment]
> @jlaitine You made a BMM350 driver as well right for PX4? Could you review this, I don't have hardware yet with a BMM350 but we're planning to.

Yes, I also offered to upstream it back then, but there was no interest I guess at the time. Our version of the driver is here, if needed for reference: https://github.com/tiiuae/px4-firmware/pull/692 .

I can check it out for sure, but not until some time next week

**Comment by jlaitine** [comment]
I started by trying it out on my BMM350 board. The driver initializes and values change when turning the mag, but the values I am getting are suspicious:

listener sensor_mag

TOPIC: sensor_mag
 sensor_mag
    timestamp: 572952315 (0.003969 seconds ago)
    timestamp_sample: 572951853 (462 us before timestamp)
    device_id: 15012873 (Type: 0xE5, I2C:1 (0x14))
    x: -14.34939
    y: -23.14031
    z: 19.13099
    temperature: 57.90157
    error_count: 0

Earth's magnetic field is supposed to be < 1 gauss, it looks like there is some very large offset on all axis. I didn't debug what that might be, but is it possible that self test internal flux was left on on the axis?

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:211]: suggestion: could just call "probe()" here

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:184]: This doesn't make sense to me,
- CHIP_ID should be readable also in suspended mode (datasheet: "Read and write of all registers is possible.")
- The actual mode should be visible in PMU_STATUS0 (PMU_CMD_STATUS_0 in datasheet) register, so should read that if one needs to find out in which mode we are in

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:231]: This looks strange. The chip should always be in SUSPEND after reset, so there is no point in putting it into suspend again. Instead of this, I'd suggest just checking in WAIT_FOR_RESET that the chip actually is in SUSPEND (as this indicates that the reset has succeeded).

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (Viliuks) [reply src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:211]: Hmm, regarding the values, it seems like I didn't check uncalibrated values.

nsh> listener vehicle_magnetometer

TOPIC: vehicle_magnetometer
 vehicle_magnetometer
    timestamp: 17710478 (0.109007 seconds ago)
    timestamp_sample: 17666390 (44088 us before timestamp)
    device_id: 15012873 (Type: 0xE5, I2C:1 (0x14))
    magnetometer_ga: [-0.22193, 0.03159, 0.24370]
    calibration_count: 3

nsh> listener sensor_mag

TOPIC: sensor_mag
 sensor_mag
    timestamp: 25610459 (0.003953 seconds ago)
    timestamp_sample: 25608889 (1570 us before timestamp)
    device_id: 15012873 (Type: 0xE5, I2C:1 (0x14))
    x: -2.21848
    y: -9.26652
    z: -12.72768
    temperature: 24.35856
    error_count: 0
    
 But i do see, that I forgot to turn off the self test for axis. Good catch, I will take a look, thanks!

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:232]: Please don't use 30ms sleeps; these pause the work queue. If you have several devices on the same i2c bus, you block those devices for the whole time, in this case > 74ms in this state. I'd suggest walking through these delays via adding states and using ScheduleDelayed().

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:303]: The datasheet says:
x = mag_x(during self-test) – mag_x(before self-test) >= 130µT
y = mag_y(during self-test) – mag_y(before self-test) >= 130µT

The code doesn't do this, it instead calculates the distance between the h and l. Both h and l are supposed to be 130µT added to opposite directions, so actually the distance should  be 260µT (?). 

I see that the code is copied from the BMM350_SensorApi, which is written by Bosch, so I don't know which one is correct. I only have tested it on handful of sensors, but for all of those the limits from the datasheet seem to work.

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:309]: Why proceed even if self test fails? I believe one should try to recover via reset instead.

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:419]: Should the pad drive actually be set as early as possible?

Comment (Viliuks) [reply src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:419]: If I'm not mistaken it should be set before returning to the normal power mode.

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:456]: This is confusing; I'd suggest something like just setting "axis_data = X_AXIS_EN | Y_AXIS_EN | Z_AXIS_EN;" and define those macros accordingly...

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:509]: This copying data araound looks strange to me, works of course. Suggest leaving out the memcpy, and just bumping the indecies below:
"uint8_t res = transfer(&cmd, 1, (uint8_t *)&mag_data, sizeof(mag_data));
...
 raw_mag_x = mag_data[2] + ((uint32_t)mag_data[3] << 8) + ((uint32_t)mag_data[4] << 16);"

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:523]: This probably gets optimized away by the compiler. However, the lsb_to_ut_degc is just a vector of constant values, so it would be cleaner to simply set it as a constant vector.

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:570]: not necessary. the function is private within the driver and never called with NULL ptr...

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:579]: I don't think that this sleep is needed... ?

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:661]: should return the read status value separate from register value. that is, better prototype would be
"int BMM350::RegisterRead(Register reg, uint8_t *val)"

This allows returning also register value "0xff", and checking whether the read succeeded. I admit that it would add some extra error checking code in the driver, but the case where i2c read would fail and the value 0xff is used would cause confusion.

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (jlaitine) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:675]: should return status value of whether the i2c write succeeded.

**Comment by jlaitine** [comment]
I added some minor comments, overall it looks quite nice to me!

I wonder if it is reasonable to poll the sensor at ODR rate, or should the default ODR be lower than 200MHz? All the older mags are polled less frequently. I don't really have a strong opinion on this, but at least I don't have use for higher frequency mag data than 50Hz or so. Just maybe it is better to set the ODR & averaging in a way that the mag produces more averaged data less frequently?

I don't think I have rights to bring review any further than commenting, so approvals need to come from maintainers...

**Comment by Viliuks** [comment]
> I added some minor comments, overall it looks quite nice to me!
> 
> I wonder if it is reasonable to poll the sensor at ODR rate, or should the default ODR be lower than 200MHz? All the older mags are polled less frequently. I don't really have a strong opinion on this, but at least I don't have use for higher frequency mag data than 50Hz or so. Just maybe it is better to set the ODR & averaging in a way that the mag produces more averaged data less frequently?
> 
> I don't think I have rights to bring review any further than commenting, so approvals need to come from maintainers...

Thank you for the review! I have updated the code accordingly. As for the big offsets I had the OTP word LSB/MSB registers mixed up.

Regarding the polling rate, I wasnt too sure too, I left it as is. If needed I can modify it to use the 50Hz polling rate as the older BMM150 magnetometer does.

**Comment by PetervdPerk-NXP** [comment]
Maybe make ODR configurable using a CLI argument and default to 50Hz.
Sometimes more is better but since it's an I2C sensor you've to share the bus with other peripherals as well.

**Comment by Viliuks** [comment]
> Maybe make ODR configurable using a CLI argument and default to 50Hz. Sometimes more is better but since it's an I2C sensor you've to share the bus with other peripherals as well.

Should I then remove the parameter settings for it? Or do I extend the functionality which would use the set parameters if no CLI arguments are passed and keep the default values for the parameters 50Hz?

**Comment by PetervdPerk-NXP** [comment]
> > Maybe make ODR configurable using a CLI argument and default to 50Hz. Sometimes more is better but since it's an I2C sensor you've to share the bus with other peripherals as well.
> 
> Should I then remove the parameter settings for it? Or do I extend the functionality which would use the set parameters if no CLI arguments are passed and keep the default values for the parameters 50Hz?

Nvm parameters are fine, boards can override using the `param set-default` command let's keep it as is.
Regarding default I would say 50Hz then.

**Comment by Viliuks** [comment]
> > > Maybe make ODR configurable using a CLI argument and default to 50Hz. Sometimes more is better but since it's an I2C sensor you've to share the bus with other peripherals as well.
> > 
> > 
> > Should I then remove the parameter settings for it? Or do I extend the functionality which would use the set parameters if no CLI arguments are passed and keep the default values for the parameters 50Hz?
> 
> Nvm parameters are fine, boards can override using the `param set-default` command let's keep it as is. Regarding default I would say 50Hz then.

Perfect, updated the module.yaml 😄

**Review by jlaitine — APPROVED** [review]
LGTM!

I also made a quick smoke test on my desk. It worked out of the box and the outputs look good.

I don't have a possibility to make any comprehensive ( or flight ) testing atm. But after this is merged, we'll replace our own driver with this one, which will then take it to the air in TII/SSRC organization.

Suggest squashing the commits before merging though

**Comment by jlaitine** [comment]
I see that this one is still pending merging, @PetervdPerk-NXP is this waiting for something?

**Comment by justas-** [comment]
We need an approval from someone with write access.

**Comment by PetervdPerk-NXP** [comment]
@justas-  Could you rebase on main and fix the CI errors?
```
../../src/drivers/magnetometer/bosch/bmm350/BMM350.cpp: In member function 'int BMM350::ReadOTPWord(uint8_t, uint16_t*)':
../../src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:643:27: error: 'lsb' may be used uninitialized in this function [-Werror=maybe-uninitialized]
  643 |    *lsb_msb = ((msb << 8) | lsb) & 0xffff;
      |               ~~~~~~~~~~~~^~~~~~
compilation terminated due to -Wfatal-errors.
```

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (PetervdPerk-NXP) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:642]: You're not checking the return code here, LSB could be corrupted.

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (PetervdPerk-NXP) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:224]: We go into PX4_ERROR here but then on line 227 we just ignore the error.

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (PetervdPerk-NXP) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:232]: Duplicate `ScheduleDelayed(10_ms);` move out of the if statement.

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (PetervdPerk-NXP) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:217]: Move the int ret allocation in this function below `const hrt_abstime now`

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (PetervdPerk-NXP) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:327]: An error ocurred but we're still processing xyzt afterwards.

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (PetervdPerk-NXP) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:507]: If writes fails during configure simply return with an error code.
We don't want to have a "half" configured sensor.

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (PetervdPerk-NXP) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:514]: See above.

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (PetervdPerk-NXP) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:521]: See above.

File: src/drivers/magnetometer/bosch/bmm350/BMM350.cpp
Comment (PetervdPerk-NXP) [inline src/drivers/magnetometer/bosch/bmm350/BMM350.cpp:527]: See above.

**Comment by PetervdPerk-NXP** [comment]
Code looking good, could you remove the merge commit, we typically rebase pull requests instead of merging them.

**Comment by Viliuks** [comment]
> Code looking good, could you remove the merge commit, we typically rebase pull requests instead of merging them.

Rebased and changed the head to remove merge commit.

**Review by PetervdPerk-NXP — APPROVED** [review]

**Review by jlaitine — APPROVED** [review]
LGTM!

**Review by dakejahl — APPROVED** [review]

**Comment by dagar** [comment]
Breaking out the configuration to parameters is fine if you're actually going to use it, but sometimes less is more. For example the estimator isn't even going to do anything with magnetometer data faster than 100 Hz. On the extreme lower end it's just going to cause problems hitting timeouts and delayed data (EKF2_MAG_DELAY) that's wildly different than typical.

A good setting would be whatever maximizes the actual data sampling (minimize potential for aliasing), while fitting comfortably on the bus at a reasonable rate (eg 100 kHz), doesn't add any additional filtering or latency in the device itself, and most importantly has actually been tested end-to-end.

**Pull request merged**
